### ISSUE TITLE
Add an option so that ASR only marks a function as intrinsic but does not change it

### ISSUE THREAD

**Issue by lucifer1004** [lfortran/lfortran#796]
Currently, the ASR tries to specify each intrinsic function. But in some cases, e.g., the Julia backend, it seems better to have an option where the ASR only marks a function intrinsic and lets the backend translate it to the equivalent function in its language.

For example, the ASR specifies `not` to `not32` or others, considering the type of the callee. However, in Julia, we can use `~x` directly and do not need to use kind-specific functions.

By adding this option (and disabling it by default so as not to break current logic), cleaner and neater code can be generated in many cases.

**Comment by certik** [comment]
We can also consider adding a specific ASR node if it makes sense for operations like these. See this document for the overall design: https://github.com/lcompilers/lpython/wiki/ASR-Design

Can you give an example of the `not` case? Let's figure it out on a case by case basis.

**Comment by lucifer1004** [comment]
For example, a simple program

```fortran
program s
   integer :: a
   a = 2
   print *, not(a)
end program
```

is translated to the following ASR

```
(TranslationUnit (SymbolTable 1 {iso_fortran_env: (IntrinsicModule lfortran_intrinsic_iso_fortran_env), lfortran_intrinsic_bit: (IntrinsicModule lfortran_intrinsic_bit), s: (Program (SymbolTable 2 {a: (Variable 2 a Local () () Default (Integer 4 []) Source Public Required .false.), not: (ExternalSymbol 2 not 4 not lfortran_intrinsic_bit [] not Private), not@not32: (ExternalSymbol 2 not@not32 4 not32 lfortran_intrinsic_bit [] not32 Private)}) s [] [(= (Var 2 a) (IntegerConstant 2 (Integer 4 [])) ()) (Print () [(FunctionCall 2 not@not32 2 not [((Var 2 a))] (Integer 4 []) () ())] () ())])}) [])
```

In which `not` is specified as the intrinsic `not@not32`. Then if we further translate this ASR into other languages, the LFortran intrinsic module will be included and a lot more code will be generated. But if we use the `not` function in the host language, then the intrinsic module is not necessary at all and the code will be cleaner.

**Comment by certik** [comment]
Is this bitwise not? We actually already have a `IntegerBitNot` ASR node. I think the LFortran frontend does not yet use it, but it should. Then this particular issue will go away.

What other cases did you discover? Let's discuss them.

**Comment by lucifer1004** [comment]
There is `IntegerBitNot` ASR node, and the Julia backend has implemented it. Problem is that in the program above, `not(a)` is not recognized as an `IntegerBitNot` node.

**Comment by lucifer1004** [comment]
Even `not(2)` is not recognized as `IntegerBitNot` either.

```fortran
program s
   print *, not(2)
end program
```

```
(TranslationUnit (SymbolTable 1 {iso_fortran_env: (IntrinsicModule lfortran_intrinsic_iso_fortran_env), lfortran_intrinsic_bit: (IntrinsicModule lfortran_intrinsic_bit), s: (Program (SymbolTable 2 {not: (ExternalSymbol 2 not 4 not lfortran_intrinsic_bit [] not Private), not@not32: (ExternalSymbol 2 not@not32 4 not32 lfortran_intrinsic_bit [] not32 Private)}) s [] [(Print () [(FunctionCall 2 not@not32 2 not [((IntegerConstant 2 (Integer 4 [])))] (Integer 4 []) (IntegerConstant -3 (Integer 4 [])) ())] () ())])}) [])
```

**Comment by certik** [comment]
> There is `IntegerBitNot` ASR node, and the Julia backend has implemented it. Problem is that in the program above, `not(a)` is not recognized as an `IntegerBitNot` node.

Yes, the LFortran frontend (AST->ASR) should start using this node, instead of linking to the library function. If you can send a PR to fix this, that would be awesome.

Regarding the "intrincis" flag, we are a bit inconsistent, but in general we do have this flag in ASR and you can check it in the backend and provide a specific implementation.

**Comment by lucifer1004** [comment]
BTW, how do we format the cpp code? I currently only format my `asr_to_julia.cpp` with my local formatter (clang-format) turned on, and when editing other files, I always turn the formatter off.

**Comment by certik** [comment]
Right now we don't, see #794.
